### DISCUSSION TITLE
Laser off between lines

### DISCUSSION THREAD

**Discussion by putt1ck** [winder/Universal-G-Code-Sender#2629] [Help] [answered]
Just started using UGS for managing a CNC with optional laser. Creating simple one word text jobs with outline mode, the laser is on throughout the movement, creating lines between letters and inside some letters e.g. "B". I can edit the gcode to set the laser off at the end of each letter/part letter, but wondering if there's a setting that would do that for me.

**Comment by breiler** [comment]
What are you using for designing/generating the gcode?

Usually the controller can be run in a "Laser mode" (https://github.com/gnea/grbl/wiki/Grbl-v1.1-Configuration#32---laser-mode-boolean) which will turn off the laser on rapid movements.

If you don't have that or enabled that setting your CAM needs to set the spindle speed output to zero between movements.

**Comment by putt1ck** [comment]
The Gcode is an export from UGS; the original "design" was done in UGS using the "Draw text" tool. The controller (FoxAlien XE Pro) has a laser switch which is turned on (it cannot switch the laser on/off without that turned on). The gcode has entries for each path sequence to turn on the laser preceding G0, in form M3  Snnnn followed by a feed speed.

In a manual edit appending an M5 to the end of path sequence turned off the laser; then the G0 entry for the next path was moved ahead of the M3 section so the laser moved to the beginning of the next letter while off. Reading the gcode (and taking into account IANAD and this is my first encounter with gcode) it seems that ensuring each G0 was preceded by an M5 and followed by the M3 section would turn off the laser at the appropriate points. I'm happy to test this theory and report back.

**Reply by breiler** [reply]
Can you post you controller settings (from the command $$).

**Reply by putt1ck** [reply]
Can confirm modifying the file as described above resulted in the laser turning off as required. Reported settings follow:
```
*** Fetching device settings
>>> $$
$0 = 10    (Step pulse time, microseconds)
$1 = 255    (Step idle delay, milliseconds)
$2 = 0    (Step pulse invert, mask)
$3 = 1    (Step direction invert, mask)
$4 = 0    (Invert step enable pin, boolean)
$5 = 0    (Invert limit pins, boolean)
$6 = 0    (Invert probe pin, boolean)
$10 = 3    (Status report options, mask)
$11 = 0.010    (Junction deviation, millimeters)
$12 = 0.002    (Arc tolerance, millimeters)
$13 = 0    (Report in inches, boolean)
$20 = 0    (Soft limits enable, boolean)
$21 = 1    (Hard limits enable, boolean)
$22 = 1    (Homing cycle enable, boolean)
$23 = 3    (Homing direction invert, mask)
$24 = 100.000    (Homing locate feed rate, mm/min)
$25 = 1000.000    (Homing search seek rate, mm/min)
$26 = 250    (Homing switch debounce delay, milliseconds)
$27 = 3.000    (Homing switch pull-off distance, millimeters)
$30 = 10000    (Maximum spindle speed, RPM)
$31 = 0    (Minimum spindle speed, RPM)
$32 = 0    (Laser-mode enable, boolean)
$100 = 400.000    (X-axis travel resolution, step/mm)
$101 = 400.000    (Y-axis travel resolution, step/mm)
$102 = 400.000    (Z-axis travel resolution, step/mm)
$110 = 5000.000    (X-axis maximum rate, mm/min)
$111 = 5000.000    (Y-axis maximum rate, mm/min)
$112 = 5000.000    (Z-axis maximum rate, mm/min)
$120 = 300.000    (X-axis acceleration, mm/sec^2)
$121 = 300.000    (Y-axis acceleration, mm/sec^2)
$122 = 300.000    (Z-axis acceleration, mm/sec^2)
$130 = 400.000    (X-axis maximum travel, millimeters)
$131 = 400.000    (Y-axis maximum travel, millimeters)
$132 = 95.000    (Z-axis maximum travel, millimeters)
ok
```

**Reply by breiler** [reply]
Yes, that is one way of doing it, turning on/off the laser output manually with M3/M5 commands. I do have some safety concerns with it though. What happens if you stop a job while the laser is firing?

A safer way is to activate $32=1 which will automatically turn on/off the laser when issuing G1, G2 and G3 commands.

**Reply by putt1ck** [reply]
I'm new here... What are the consequences of enabling that and leaving it enabled when the laser is swapped out for the spindle?

**Reply by breiler** [reply, marked as answer]
That will turn off the spindle between rapid moves, so that could have the undesired effect that the spindle would not have time to spin up to the proper speed before it would engage the material. So that I would advice against using a spindle in laser mode.

I could probably add an option to use M3/M5 for lasers.

**Reply by putt1ck** [reply]
Maybe it could be a macro
